_specifications/specification-3-16.md

```diff
@@ -2779,6 +2779,31 @@ _Response_:
 * result: `any` \| `null`
 * error: code and message set in case an exception happens during the request.
 
+#### <a href="#client_executeCommand" name="client_executeCommand" class="anchor">Execute a client command(:leftwards_arrow_with_hook:)</a>
+
+The `client/executeCommand` request is similar to `workspace/executeCommand` but is sent from the server to the client, to trigger command execution on the client. 
+
+_Client Capability_:
+* property path (optional): `client.executeCommand`
+* property type: `ClientExecuteCommandClientCapabilities` defined as follows:
+
+```typescript
+export interface ClientExecuteCommandClientCapabilities {
+	/**
+	 * The client supports message 'client/executeCommand'.
+	 */
+	supported: boolean;
```

Comment:
I get that you're trying get something out the door but this needs many more options for clients. Given just this state of the PR a server doesn't know whether a command can succeed or not.

Just to give you some ideas:

- the client should be able to communicate dynamically to the server about possible new commands that it knows about, similar to workspace/didChangeConfiguration.
- It should be possible for the server to get a list of known command names from the client, similar to workspace/configuration

At this point the server can only do client/executeCommand and hope for the best.

Reply:
I don't think we should require more options at this point. The current proposal is good and there is room for evolution (such as listing commands) in the future. However, we don't really have a strong need to pass a static list of commands at that point, just like symmetrically, there is no strong need for clients to know in advance the list of server commands it can invoke.
The command pattern is often use for client or LS specific integrations, that we cannot really standardize in LSP because they're not generic enough.
I think it's better to mimic the current `workspace/executeCommand` action (and related capabilities), it seems more consistent.

> At this point the server can only do client/executeCommand and hope for the best.

It's the same for `workspace/executeCommand`, and yet this is working well ;)

Reply:
> It's the same for workspace/executeCommand, and yet this is working well ;)

Not quite. If you're lucky, you can read a readme file explaining what the arguments of a command can be. If you're unlucky, you have to dig through the server source code to figure out what a command can do.

Reply:
The configure-less way of invoking commands is via code actions, which works well, I agree.

Reply:
I'm not against adding more options for things such as 'command discovery' or making the spec more precise about mechanism by which commands are made to 'exist' on the client side.

However, I wanted to explicitly avoid getting this become yet another ticket dragged out into a lengthy discussion with no actionable outcome. 

To me the questions:

1. how does a server request a client to execute a command
2. how does a server know that a client can execute a specific (set of) commands
3. how are client-side commands defined

So I explicitly wanted limit the scope of this ticket to be just about 1. The other questions can be answered independently and in fact there are already some tickets open with long discussions.

> Given just this state of the PR a server doesn't know whether a command can succeed or not.

That is true, but at least we know how to request execution in a standardised way. To me that is progress. 

And we can assume an error comes back if the execution failed (I agree with your other comments that we can perhaps make some of the error cases more explicit in the spec).

Reply:
No worries. I am just some guy maintaining a client and weary of opening pandora's box with client/executeCommand. We must have a follow-up PR that defines how command discovery works, and we must standardize some "common" commands.

Reply:
I don't really disagree with you. I just don't want to go there yet.  Yes, I could imagine some options added to capability object in the future such as 'commands: [...]' for a list of 'statically' known commands. Or a property 'commandDisovery: boolean' that indicates there is some extra protocol for command discovery. etc. But this sort of thing could be arbitrarily complex and with lots of people wanting something a little different from it. So I just don't think we should/can tackle all of these at once right now. The more we can separate this into smaller questions/issues, the more chance we have of actually making progress in small increments.